scrapy/core/http2/stream.py

```diff
@@ -0,0 +1,384 @@
+        )
+
+    def receive_headers(self, headers):
+        for name, value in headers:
+            self._response['headers'][name] = value
+
+        # Check if we exceed the allowed max data size which can be received
+        expected_size = int(self._response['headers'].get(b'Content-Length', -1))
+        if self._download_maxsize and expected_size > self._download_maxsize:
+            self.reset_stream(StreamCloseReason.MAXSIZE_EXCEEDED)
+            return
+
+        if self._download_warnsize and expected_size > self._download_warnsize:
+            warning_msg = ("Expected response size ({size}) larger than ",
+                           "download warn size ({warnsize}) in request {request}.")
+            warning_args = {
+                'size': expected_size, 'warnsize': self._download_warnsize,
+                'request': self._request
+            }
+            LOGGER.warning(warning_msg, warning_args)
+
+    def reset_stream(self, reason=StreamCloseReason.RESET):
+        """Close this stream by sending a RST_FRAME to the remote peer"""
+        if self.stream_closed_local:
+            raise StreamClosedError(self.stream_id)
+
+        self.stream_closed_local = True
+        self._conn.reset_stream(self.stream_id, ErrorCodes.REFUSED_STREAM)
+        self._write_to_transport()
+        self.close(reason)
+
+    def _is_data_lost(self) -> bool:
+        assert self.stream_closed_server
+
+        expected_size = self._response['flow_controlled_size']
+        received_body_size = int(self._response['headers'][b'Content-Length'])
+
+        return expected_size != received_body_size
+
+    def close(self, reason: StreamCloseReason, failure=None):
+        """Based on the reason sent we will handle each case.
+        """
+        if self.stream_closed_server:
+            raise StreamClosedError(self.stream_id)
+
+        self._cb_close(self.stream_id)
+        self.stream_closed_server = True
+
+        # Do nothing if the response deferred was cancelled
+        flags = None
+        if b'Content-Length' not in self._response['headers']:
+            # Missing Content-Length - PotentialDataLoss
+            flags = ['partial']
+        elif self._is_data_lost():
+            if self._fail_on_dataloss:
+                self._deferred_response.errback(ResponseFailed([Failure()]))
+                return
+            else:
+                flags = ['dataloss']
+
+        if reason is StreamCloseReason.ENDED:
+            self._fire_response_deferred(flags)
+
+        # Stream was abruptly ended here
+        elif reason is StreamCloseReason.CANCELLED:
+            # Client has cancelled the request. Remove all the data
+            # received and fire the response deferred with no flags set
+            self._response['body'].truncate(0)
+            self._response['headers'].clear()
+            self._fire_response_deferred()
+
+        elif reason in (StreamCloseReason.RESET | StreamCloseReason.CONNECTION_LOST):
```

Comment:
```suggestion
        elif reason in (StreamCloseReason.RESET, StreamCloseReason.CONNECTION_LOST):
```
Is there a reason to be using a bitwise or? I found it a bit confusing, I had to look into the docs to understand the subtle fact that [`enum.IntFlag`](https://docs.python.org/3/library/enum.html#enum.IntFlag) handles this (`1 in (1 | 2)` explodes as expected). I believe doing a "regular" tuple check is simpler and more intuitive. It would also allow to define `StreamCloseReason` as a subclass of `enum.Enum` instead.
Let me know if I'm missing something.

Reply:
You're right. I didn't get the idea of using a `in` tuple check earlier so went with `IntFlag`. Changed definition to
```python
class StreamCloseReason(Enum):
```